File: backend/src/Alife.Domain/Entities/EventVenue.cs
using Alife.Domain.Enums;

namespace Alife.Domain.Entities;

/// <summary>A reusable venue managed by a group. Event/session display text is not authoritative venue data.</summary>
public sealed class EventVenue
{
    public Guid Id { get; set; }
    public Guid ManagingGroupId { get; set; }
    public string NameEn { get; set; } = string.Empty;
    public string NameZh { get; set; } = string.Empty;
    public string AddressEn { get; set; } = string.Empty;
    public string AddressZh { get; set; } = string.Empty;
    public int Capacity { get; set; }
    public bool IsActive { get; set; } = true;
    public Guid ConcurrencyToken { get; set; } = Guid.NewGuid();
    public Guid CreatedByMemberId { get; set; }
    public DateTime CreatedUtc { get; set; }
    public DateTime UpdatedUtc { get; set; }

    public Group ManagingGroup { get; set; } = null!;
    public Member CreatedByMember { get; set; } = null!;
    public ICollection<EventVenueReservation> Reservations { get; set; } = [];
}

/// <summary>
/// An auditable venue booking. EventId is always present for ownership and authorisation;
/// EventOccurrenceId narrows the booking to one occurrence when supplied.
/// </summary>
public sealed class EventVenueReservation
{
    public Guid Id { get; set; }
    public Guid VenueId { get; set; }
    public Guid EventId { get; set; }
    public Guid? EventOccurrenceId { get; set; }
    public DateTime StartUtc { get; set; }
    public DateTime EndUtc { get; set; }
    public int RequiredCapacity { get; set; }
    public EventVenueReservationStatus Status { get; set; } = EventVenueReservationStatus.Confirmed;
    public Guid ReservedByMemberId { get; set; }
    public Guid? ReleasedByMemberId { get; set; }
    public DateTime? ReleasedUtc { get; set; }
    public Guid ConcurrencyToken { get; set; } = Guid.NewGuid();
    public DateTime CreatedUtc { get; set; }
    public DateTime UpdatedUtc { get; set; }

    public EventVenue Venue { get; set; } = null!;
    public GroupEvent Event { get; set; } = null!;
    public EventOccurrence? EventOccurrence { get; set; }
    public Member ReservedByMember { get; set; } = null!;
    public Member? ReleasedByMember { get; set; }
}

File: cloudflare/alife-app/src/types/eventVenue.ts
import type { LocalizedText } from './eventComposition'

export type EventVenue = {
  id: string
  managingGroupId: string
  name: LocalizedText
  address: LocalizedText
  capacity: number
  isActive: boolean
  eTag: string
  createdUtc: string
  updatedUtc: string
}

export type EventVenueReservationStatus = 'confirmed' | 'released'

export type EventVenueReservation = {
  id: string
  venueId: string
  eventId: string
  eventOccurrenceId?: string | null
  venueName: LocalizedText
  venueCapacity: number
  startUtc: string
  endUtc: string
  requiredCapacity: number
  status: EventVenueReservationStatus
  reservedByMemberId: string
  releasedByMemberId?: string | null
  releasedUtc?: string | null
  eTag: string
  createdUtc: string
  updatedUtc: string
}

export type EventVenueConflict = {
  venueId: string
  venueName: LocalizedText
  startUtc: string
  endUtc: string
}

export type EventVenueWorkspace = {
  eventId: string
  managingGroupId: string
  venues: EventVenue[]
  reservations: EventVenueReservation[]
  conflicts: EventVenueConflict[]
  readiness: {
    capacitySufficient: boolean
    bookingsConfirmed: boolean
    conflictsResolved: boolean
    blockers: LocalizedText[]
  }
  canManage: boolean
  canManageCatalogue: boolean
  legacySessionPlacePreserved: boolean
}

export type SaveEventVenueRequest = {
  name: LocalizedText
  address: LocalizedText
  capacity: number
  isActive: boolean
}

export type ReserveEventVenueRequest = {
  venueId: string
  eventOccurrenceId?: string | null
  startUtc: string
  endUtc: string
  requiredCapacity: number
}
